Map/src/main/java/org/Marta/MyMap.java

```diff
@@ -0,0 +1,173 @@
+                continue;
+            }
+            for (Entry<K, V> entry : bucket) {
+                int hash = hash(entry.key);
+                if (table[hash] == null) {
+                    table[hash] = new LinkedList<>();
+                }
+                List<Entry<K, V>> newBucket = table[hash];
+                newBucket.add(entry);
+            }
+        }
+    }
+
+    @Override
+    public boolean put(K key, V value) {
+        int hash = hash(key);
+        if (table[hash] == null) {
+            table[hash] = new LinkedList<>();
+        }
+        boolean keyExist = false;
+        List<Entry<K, V>> bucket = table[hash];
+
+        for (Entry<K, V> entry : bucket) {
+            if (key.equals(entry.key)) {
+                entry.value = value;
+                keyExist = true;
+                break;
+            }
+        }
+
+        if (!keyExist) {
+            bucket.add(new Entry<>(key, value));
+            size++;
+        }
+
+        if (size == threshold) {
+            resize();
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean containsKey(K key) {
+        return getEntry(key) != null;
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        if (!isEmpty()) {
+            for (List<Entry<K, V>> bucket : table) {
+                if (bucket == null) {
+                    continue;
+                }
+                for (Entry<K, V> entry : bucket) {
+                    if (Objects.equals(value, entry.value)) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public V remove(K key) {
+        List<Entry<K, V>> bucket = table[hash(key)];
+        if (bucket == null) {
+            return null;
+        }
+
+        Iterator<Entry<K, V>> bucketIterator = bucket.iterator();
+        V oldValue = null;
+        while (bucketIterator.hasNext()) {
+            Entry<K, V> entry = bucketIterator.next();
+            if (key.equals(entry.key)) {
+                oldValue = entry.value;
+                bucketIterator.remove();
+                size--;
+                break;
+            }
+            if (isEmpty()) {
+                return null;
+            }
+        }
+
+        return oldValue;
+    }
+
+    private Entry<K, V> getEntry(K key) {
+        List<Entry<K, V>> bucket = table[hash(key)];
+        if (bucket != null) {
+            for (Entry<K, V> entry : bucket) {
+                if((key == null && entry.key == null) || Objects.equals(key, entry.key))
+                    return entry;
+                }
+            }
+        return null;
+    }
+
+    @Override
+    public V get(K key) {
+        Entry<K, V> entry = getEntry(key);
+        if (entry == null) {
+            return null;
+        }
+        return entry.value;
+    }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    //for tests
```

Comment:
Wszystko super, ale to mi sie nie podoba. Nigdy nie rób metod, które są tylko dla testów.

![image](https://user-images.githubusercontent.com/17989854/120273731-2e7de500-c2af-11eb-91e1-9d0a7a0d3411.png)


więc te metody 'for tests' do wywalenia - testujesz tylko i wyłącznie publiczne.

Reply:
wiem, że czasami ciężko to zrobić no ale nikt nie powiedział, że testy sa łatwe ;)